pkg/svcat/service-catalog/instance.go

```diff
@@ -201,6 +201,27 @@ func (sdk *SDK) Deprovision(namespace, instanceName string) error {
 	return nil
 }
 
+// UpdateInstance updates an instance of a service class and plan.
+func (sdk *SDK) UpdateInstance(namespace, instanceName, planName string,
+	params interface{}, secrets map[string]string) (*v1beta1.ServiceInstance, error) {
+
+	inst, err := sdk.RetrieveInstance(namespace, instanceName)
+	if err != nil {
+		return nil, err
+	}
+
+	inst.Spec.PlanReference.ClusterServicePlanExternalName = planName
```

Comment:
The `set` command should use the PATCH verb, not PUT (i.e. update). This allows the command to modify a single property without accidentally colliding with another action that may be modifying the resource as well.

Once you move to PATCH, you will be able to remove a bunch of the logic around populating parms, plan, secrets, etc based on existing values, and only modify what the user requested.

Reply:
I understand what you're saying.

But still, my thinking was that I would still need to make three *if* checks to see if **ClusterServicePlanExternalName**, **Parameters** or **ParametersFrom** is different from that which is persisted already (so I know which one to update). This would be more idiomatic of a PATCH request, yes, but would require additional checks for each property.

That is why in the *instance/set_cmd.go* file, where the implementation of the set instance command resides, I made the **configureCmdProperties** function, in order to detect if some of the flags was not provided, so I can retrieve and assign the value persisted in the instance already. This would make the logic that we're now discussing lighter, since we would not need to check for each property's value here (this has already been taken care of in the **configureCmdProperties** function).

Another thing that's concerning for me would be that if I add those checks here, it would be like checking for the same thing twice (since the **configureCmdProperties** function does this also). You might say "Why not scrap the **configureCmdProperties** function all together and make all the checks in the function we're now discussing?". Well, if the user doesn't provide a value for any of the optional flags, the **set_cmd.go** logic would have to pass empty values to this module right here. How would we distinguish when there are no parameters truly passed and when there's a request to override the properties of a instance to empty? This is why I think the current setup is the most optimal for the goal we're trying to achieve. Just update all the parameters, knowing that the **set_cmd.go** would take care of passing in the correct values.

What do you think? 😌

Reply:
Sorry for the lack of a response, Carolyn's been busy with other stuff. 

First off, I dont think configureCmdProperties in the cmd layer is the way to go. I think the cmd layer shouldn't have to know anything about the actual interaction with the cluster, as that's the job of the pkg layer. So, if we moved configureCmdProperties to the pkg layer, this would allow us to make use of PATCH instead.

As for the blank vs not provided problem, couldn't it we use and empty byte array `[]byte{}` for not provided, and an empty string byte array for `[]byte("") for intentionally-left-empty?

Reply:
Any update on this?

Reply:
@jberkhahn Okay, so I decided to calling the Patch method instead of Update on line 215 and really this whole function now looks something like this:

```go
// UpdateInstance updates an instance of a service class and plan.
func (sdk *SDK) UpdateInstance(namespace, instanceName, planName string,
	params interface{}, secrets map[string]string) (*v1beta1.ServiceInstance, error) {
	instance := &v1beta1.ServiceInstance{
		Spec: v1beta1.ServiceInstanceSpec{
			PlanReference: v1beta1.PlanReference{
				ClusterServicePlanExternalName:  planName,
			},
			Parameters: BuildParameters(params),
			ParametersFrom: BuildParametersFrom(secrets),
		},
	}

	instanceJSON, err := json.Marshal(instance)
	if err != nil {
		return nil, fmt.Errorf("unable to marshal instance (%s)", err)
	}

	result, err := sdk.ServiceCatalog().ServiceInstances(namespace).Patch(instanceName, types.JSONPatchType, instanceJSON)
	if err != nil {
		return nil, fmt.Errorf("update instance request failed (%s)", err)
	}
	return result, nil
}
```

But now I get `Error: update instance request failed (json: cannot unmarshal object into Go value of type jsonpatch.Patch)` after the Patch call.

I'm not sure if I'm calling the Patch method correctly, since it has never been used in the Service Catalog project - any suggestions?

Reply:
@jberkhahn Can you assist with this?

Reply:
You need to structure the JSON correctly.

You're just chucking the instance object into JSON and trying to patch with that. That's not how a JSON patch works. See this issue: https://github.com/kubernetes/kubernetes/issues/68861 and some docs on how JSON patching works: http://jsonpatch.com/

Reply:
@jberkhahn Thanks for the heads up regarding JSON-Patch!

I played around with the implementation quite a lot trying to figure out how to make this work in more in a Patch manner, rather than Update like it is right now. I've reached a road block with various types of implementation approaches. Here are benefits & drawbacks to each of the Patch approaches:

**JSON-Patch approach:**

Benefits:
- No need to provide ExternalID property

Drawbacks:
- Need to construct JSON path manually for each patched property as part of each JSON patch operation (we can’t just work with the default Go structs that are provided by the Service Catalog)
- The OSB spec requires that for update instance operations the service_id is provided as part of the request body; in the cases where a plan update is performed, the service_id will be included under-the-hood (we don’t need to worry about that); but for cases where the plan for an instance is not updated we need to fetch the original service instance and artificially add a JSON patch operation which overwrites the same old plan, just so we can have the service_id added to the payload
- There’s no trivial way to insert parameters which are part of a secret (ParametersFromSource); we would most likely have to fetch the secret’s data and then find the specific key of interest so we can place it as a parameter for the instance.


**Merge/StrategicMerge Patch approach:**

Benefits:
- No need to construct manually JSON patch operations
- We use the default Go structs that are provided by the Service Catalog (ServiceInstance)
- We can handle parameters from secrets more gracefully (because we can use the Go structs directly)

Drawbacks:
- Need to provide ExternalID property during the update operation, which will require fetching the original instance beforehand.
- As mentioned above regarding the requirement to provide service_id with the request payload, we would most likely have to again provide the original plan if a new one is not given
- A patch of an instance’s parameters is interpreted as adding another parameter to the instance, instead of replacing all previous parameters, which effectively makes it impossible to remove old parameters.

**For both approaches:**
The unfortunate thing is that in either approach, we need to make two DB calls (which won’t be in a single transaction, potential for race condition) - one to fetch the original instance and one to perform the patch.

Seems like this PR isn't as trivial as I first thought and I really don't have much spare time to work on it. So I don't mind someone taking this over if the current implementation doesn't suffice you.